armo un ELT en jupiter para probar el proceso y luego migrarlo al .py

importamos las librerias, puede que haya de más pero luego ya las iremos depurando

In [2]:
import os
import pandas as pd
from pandas.io import sql
import numpy as np
import chardet
from pathlib import Path as pl
import pyarrow
import mysql
import re
import json
import datetime


importamos las funciones que vamos creando en paralelo para el ETL

In [3]:
from funciones_etl import *

creamos las variables con las paths_ vamos a respetar el orden alfabético de los archivos durante el EDA y ETL

In [4]:
path_amazon="Datasets\\amazon_prime_titles.csv"
path_disney="Datasets\\disney_plus_titles.csv"
path_hulu="Datasets\\hulu_titles.csv"
path_netflix="Datasets\\netflix_titles.json"

Se abren los archivos con la función de apertura creada y luego arrancamos el EDA


In [5]:
df_amazon=abriendo_archivos(path_amazon)
df_disney=abriendo_archivos(path_disney)
df_hulu=abriendo_archivos(path_hulu)
df_netflix=abriendo_archivos(path_netflix)

DESARROLLO DEL EDA:
- 1° -  se abrieron los archivos csv en note pad++ para ver que tipo de separación tenía y encoding 
con visual studio code, con pandas y python: 
- 2° - se corre un sample(20) varias veces cada uno para descrubrir que tipos de datos hay
- 3° - con ,info se analizan cuantos registros y cuantas columnas tienen
- 4° - se analizan los valores faltantes
- 5° - se analizan los duplicados de de los IP (count. distinct de la columna)
- 6° - aqui se debería realizar un informe al cliente y tomar decisiones sobre la estrategia de ETL


In [6]:
#"s1896" in df_amazon.show_id.values

analisis global con .sample, se ven los nombres de las columnas, puede iterarse y se ve el general de que tipo de datos se trata

In [19]:
df_amazon.sample(20)

show_id     type                                              title  \
4754   s4755    Movie                                   The Crystal Ball   
8258   s8259    Movie                                      Stuart Little   
7698   s7699    Movie                                The Habit of Beauty   
6450   s6451    Movie                                          Purgatory   
2417   s2418    Movie                              Frontier Pony Express   
3858   s3859    Movie                                          Gully Boy   
7595   s7596    Movie                                     The Waste Land   
5571   s5572    Movie  Little Baby Bum - Learn Colors, Shapes and Num...   
4640   s4641    Movie                                             Bhoomi   
6772   s6773    Movie                                   Marital Problems   
7752   s7753    Movie                                Conjuring the Genie   
602     s603  TV Show                                     The Invisibles   
6203   s6204    Movie                                         The Hunted   
6869   s6870    Movie                                  Angel in Training   
1735   s1736    Movie                                              Mario   
9619   s9620    Movie                                    The Shadow Side   
4483   s4484    Movie                          Ninu Veedani Needanu Nene   
9437   s9438    Movie                                      Open Marriage   
7489   s7490  TV Show                            LOL - Enga Siri Paappom   
3840   s3841    Movie                   Jeff Dunham: Arguing with Myself   

                        director  \
4754              Elliott Nugent   
8258                 Rob Minkoff   
7698              Mirko Pincelli   
6450                   Ami Artzi   
2417                 Joseph Kane   
3858                 Zoya Akhtar   
7595                   WANG MENG   
5571               Cannis Holder   
4640                 Omung Kumar   
6772                  Dia Taylor   
7752               Scott Jeffrey   
602                          NaN   
6203                Josh Stewart   
6869  Chick Vennera, Gary Graver   
1735               Marcel Gisler   
9619                 Max Coronel   
4483               Caarthik Raju   
9437                   Sam Irvin   
7489                         NaN   
3840             Manny Rodriguez   

                                                   cast         country  \
4754       Ray Milland, Paulette Goddard, Gladys George             NaN   
8258  Geena Davis, Michael J. Fox, Hugh Laurie, Jona...             NaN   
7698  Vincenzo Amato, Francesca Neri, Noel Clarke, N...             NaN   
6450  Tanya Roberts, Julie Pop, Hal Orlandini, Adrie...             NaN   
2417  Roy Rogers, Lynne Roberts, Raymond Hatton, Edw...             NaN   
3858  Ranveer Singh, Alia Bhatt, Kalki Koechlin, Sid...             NaN   
7595                           WANG MENG, Jiao Zhonghui             NaN   
5571                           Grace Tamayo, Erin Webbs             NaN   
4640  Sanjay Dutt, Aditi Rao Hydari, Sharad Kelkar, ...           India   
6772  Nick Capper, Callum Gault, Aleis Duffy, Neil G...             NaN   
7752  Sarah T. Cohen, Megan Purvis, Barbara Dabson, ...  United Kingdom   
602          Anthony Head, Warren Clarke, Jenny Agutter             NaN   
6203  Josh Stewart, Skipp Sudduth, Katherine Von Til...             NaN   
6869  Laila Dagher, Danielle Pessis, Veronica Allen,...             NaN   
1735         Max Hubacher, Aaron Altaras, Jessy Moravec             NaN   
9619           Clara Kovacic, Germán Baudino, Gon Spina             NaN   
4483  Sundeep Kishan, Anya Singh, Murali Sharma, Ven...             NaN   
9437  Tilky Jones, Nikki Leigh, Kelly Dowdle, Jason ...             NaN   
7489  Padmashri Vivek,  Shiva,  Abishek,  Baggy,  Ha...             NaN   
3840                                        Jeff Dunham             NaN   

     date_added  release_year   rating  duration  \
4754        NaN        

In [8]:
df_amazon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9668 entries, 0 to 9667
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       9668 non-null   object
 1   type          9668 non-null   object
 2   title         9668 non-null   object
 3   director      7586 non-null   object
 4   cast          8435 non-null   object
 5   country       672 non-null    object
 6   date_added    155 non-null    object
 7   release_year  9668 non-null   int64 
 8   rating        9331 non-null   object
 9   duration      9668 non-null   object
 10  listed_in     9668 non-null   object
 11  description   9668 non-null   object
dtypes: int64(1), object(11)
memory usage: 906.5+ KB


In [9]:
df_amazon.isnull().sum()

show_id            0
type               0
title              0
director        2082
cast            1233
country         8996
date_added      9513
release_year       0
rating           337
duration           0
listed_in          0
description        0
dtype: int64

In [20]:
df_disney.sample(20)

show_id     type                                    title  \
773     s774    Movie                             Can of Worms   
587     s588    Movie                               Sea Scouts   
879     s880  TV Show                                  Encore!   
963     s964    Movie                       Inspector Gadget 2   
303     s304    Movie                                You Again   
1023   s1024    Movie                          Lonesome Ghosts   
1067   s1068    Movie                       Meet the Robinsons   
1109   s1110    Movie                         Oliver & Company   
780     s781    Movie             Cars Toon: Heavy Metal Mater   
987     s988  TV Show                               Kickin' It   
598     s599  TV Show                              Be Our Chef   
491     s492    Movie   Marvel Studios' Avengers: Infinity War   
290     s291    Movie                      Myth: A Frozen Tale   
882     s883    Movie    Expedition Mars: Spirit & Opportunity   
1248   s1249    Movie                           The Aristocats   
839     s840  TV Show  Disney Tangled: The Series - Short Cuts   
1328   s1329    Movie                  The Rescuers Down Under   
827     s828  TV Show            Disney Gravity Falls (Shorts)   
1375   s1376    Movie                    Toby Tortoise Returns   
96       s97  TV Show                          Dog: Impossible   

                        director  \
773               Paul Schneider   
587                   Dick Lundy   
879                          NaN   
963                    Alex Zamm   
303                 Andy Fickman   
1023                Burt Gillett   
1067            Stephen Anderson   
1109             George Scribner   
780                John Lasseter   
987                          NaN   
598                          NaN   
491     Anthony Russo, Joe Russo   
290                  Jeff Gipson   
882                   Mark Davis   
1248         Wolfgang Reitherman   
839                          NaN   
1328  Hendel Butoy, Mike Gabriel   
827                          NaN   
1375             Wilfred Jackson   
96                           NaN   

                                                   cast  \
773   Michael Shulman, Erika Christensen, Adam Wylie...   
587                                                 NaN   
879   Kristen Bell, Roger Love, Susan Egan, Ali Stro...   
963   French Stewart, Elaine Hendrix, Caitlin Wachs,...   
303   Kristen Bell, Sigourney Weaver, Jamie Lee Curt...   
1023                                                NaN   
1067  Angela Bassett, Daniel Hansen, Jordan Fry, Mat...   
1109  Joey Lawrence, Billy Joel, Cheech Marin, Richa...   
780                Larry the Cable Guy , Keith Ferguson   
987   Leo Howard, Dylan Snyder, Mateo Arias, Olivia ...   
598                                       Angela Kinsey   
491   Robert Downey Jr., Chris Hemsworth, Mark Ruffa...   
290                                    Evan Rachel Wood   
882                                           Bray Poor   
1248  Phil Harris, Eva Gabor, Sterling Holloway, Sca...   
839                                                 NaN   
1328  Bob Newhart, Eva Gabor, John Candy, Tristan Ro...   
827                                                 NaN   
1375                                                NaN   
96                                         Matt Beisner   

                    country          date_added  release_year rating  \
773   United States, Canada   November 12, 2019          1999  TV-PG   
587           United States       April 3, 2020          1939  TV-PG   
879           United States   November 12, 2019          2019  TV-14   
963           United States   November 12, 2019          2003      G   
303           United States   February 12, 2021          2010     PG   
1023          United States   November 12, 2019          1937  TV-Y7   
1067          United States   November 12, 2019          2007      G   
1109          United States   November 12, 20

In [11]:
df_disney.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1450 entries, 0 to 1449
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       1450 non-null   object
 1   type          1450 non-null   object
 2   title         1450 non-null   object
 3   director      977 non-null    object
 4   cast          1260 non-null   object
 5   country       1231 non-null   object
 6   date_added    1447 non-null   object
 7   release_year  1450 non-null   int64 
 8   rating        1447 non-null   object
 9   duration      1450 non-null   object
 10  listed_in     1450 non-null   object
 11  description   1450 non-null   object
dtypes: int64(1), object(11)
memory usage: 136.1+ KB


In [12]:
df_disney.isnull().sum()

show_id           0
type              0
title             0
director        473
cast            190
country         219
date_added        3
release_year      0
rating            3
duration          0
listed_in         0
description       0
dtype: int64

In [13]:
df_hulu.sample(20)

show_id     type                                   title director  cast  \
1846   s1847  TV Show                                  Magi 2      NaN   NaN   
134     s135  TV Show                              Station 19      NaN   NaN   
2760   s2761  TV Show                 Mobile Suit Gundam Wing      NaN   NaN   
2025   s2026    Movie                                  Wounds      NaN   NaN   
1998   s1999    Movie                   Framing John DeLorean      NaN   NaN   
1579   s1580    Movie                               Buffaloed      NaN   NaN   
1954   s1955    Movie                             ID: INVADED      NaN   NaN   
2772   s2773  TV Show                                 Chowder      NaN   NaN   
2109   s2110    Movie       The Amazing Johnathan Documentary      NaN   NaN   
535     s536    Movie                        Personal Shopper      NaN   NaN   
2526   s2527  TV Show                         Deadliest Catch      NaN   NaN   
2189   s2190  TV Show                   Blinded by the Lights      NaN   NaN   
870     s871  TV Show                             Mayans M.C.      NaN   NaN   
579     s580  TV Show     When Nature Calls With Helen Mirren      NaN   NaN   
2403   s2404  TV Show                              The League      NaN   NaN   
110     s111    Movie                                 Peeples      NaN   NaN   
1560   s1561  TV Show             The Day I Picked My Parents      NaN   NaN   
1115   s1116    Movie              Once Upon a Time in Venice      NaN   NaN   
1994   s1995  TV Show  Phantasy Star Online 2: Episode Oracle      NaN   NaN   
834     s835  TV Show                    Say Yes to the Dress      NaN   NaN   

                            country         date_added  release_year  \
1846                            NaN   January 31, 2020          2012   
134                   United States    October 1, 2021          2018   
2760           Japan, United States  February 20, 2017          1995   
2025  United States, United Kingdom   October 18, 2019          2019   
1998                  United States   November 5, 2019          2019   
1579                  United States      June 18, 2020          2019   
1954                          Japan  December 15, 2019          2019   
2772           United States, China    January 1, 2017          2014   
2109                  United States    August 16, 2019          2019   
535                             NaN       July 1, 2021          2016   
2526                            NaN     March 15, 2018          2014   
2189                            NaN       June 1, 2019          2019   
870           Mexico, United States     March 17, 2021          2018   
579                             NaN      June 25, 2021          2021   
2403                  United States    October 1, 2018          2009   
110                             NaN    October 1, 2021          2013   
1560                  United States       July 1, 2020          2019   
1115                  United States  December 15, 2020          2017   
1994                          Japan   November 6, 2019          2019   
834                   United States      April 1, 2021          2014   

        rating   duration                                     listed_in  \
1846     TV-14   1 Season                      Action, Adventure, Anime   
134      TV-14  5 Seasons                                 Action, Drama   
2760     TV-MA   1 Season         Anime, International, Science Fiction   
2025         R     95 min                              Horror, Thriller   
1998   109 min        NaN                                 Documentaries   
1579    95 min        NaN                                        Comedy   
1954  1 Season        NaN                           Anime, Crime, Drama   
2772     TV-Y7  3 Seasons                        Cartoons, Comedy, Kids   
2109     TV-MA     91 min                                 Documentaries   
535          R    105 min                        Drama, Horro

In [14]:
df_hulu.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3073 entries, 0 to 3072
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   show_id       3073 non-null   object 
 1   type          3073 non-null   object 
 2   title         3073 non-null   object 
 3   director      3 non-null      object 
 4   cast          0 non-null      float64
 5   country       1620 non-null   object 
 6   date_added    3045 non-null   object 
 7   release_year  3073 non-null   int64  
 8   rating        2553 non-null   object 
 9   duration      2594 non-null   object 
 10  listed_in     3073 non-null   object 
 11  description   3069 non-null   object 
dtypes: float64(1), int64(1), object(10)
memory usage: 288.2+ KB


In [15]:
df_hulu.isnull().sum()

show_id            0
type               0
title              0
director        3070
cast            3073
country         1453
date_added        28
release_year       0
rating           520
duration         479
listed_in          0
description        4
dtype: int64

In [16]:
df_netflix.sample(20)

show_id     type                                title  \
4862   s4863  TV Show         Saint Seiya: The Lost Canvas   
8702   s8703    Movie                          Waterschool   
3011   s3012  TV Show                    Hip-Hop Evolution   
3787   s3788  TV Show                            Bad Blood   
624     s625    Movie                           Prime Time   
8211   s8212    Movie                       The Black Room   
226     s227  TV Show  Post Mortem: No One Dies in Skarnes   
6161   s6162  TV Show                       Ancient Aliens   
2022   s2023  TV Show               Julie and the Phantoms   
6200   s6201    Movie               Avengers: Infinity War   
3178   s3179    Movie                      Wives on Strike   
3210   s3211    Movie                       I Lost My Body   
6112   s6113    Movie           Alien Contact: Outer Space   
2954   s2955    Movie                               Thambi   
8191   s8192    Movie                    The Animal People   
6545   s6546    Movie                  Cutie and the Boxer   
6686   s6687    Movie                         Elstree 1976   
1186   s1187  TV Show                     B: The Beginning   
3019   s3020    Movie                             NiNoKuni   
6216   s6217  TV Show           Badalte Rishton Ki Dastaan   

                          director  \
4862                          None   
8702                  Tiffanie Hsu   
3011                          None   
3787                          None   
624                   Jakub Piątek   
8211                Rolfe Kanefsky   
226                           None   
6161                          None   
2022                  Kenny Ortega   
6200      Anthony Russo, Joe Russo   
3178                   Omoni Oboli   
3210                 Jérémy Clapin   
6112               J. Michael Long   
2954                 Jeethu Joseph   
8191  Denis Hennelly, Casey Suchan   
6545           Zachary Heinzerling   
6686                     Jon Spira   
1186                          None   
3019              Yoshiyuki Momose   
6216                          None   

                                                   cast  \
4862  Tetsuya Kakihara, Hiro Shimono, Aya Hirano, At...   
8702                                               None   
3011                                       Shad Kabango   
3787  Anthony LaPaglia, Kim Coates, Enrico Colantoni...   
624   Bartosz Bielenia, Magdalena Popławska, Andrzej...   
8211  Natasha Henstridge, Lukas Hassel, Lin Shaye, D...   
226   Kathrine Thorborg Johansen, Elias Holmen Søren...   
6161                                  Robert Clotworthy   
2022  Madison Reyes, Charlie Gillespie, Owen Patrick...   
6200  Robert Downey Jr., Josh Brolin, Mark Ruffalo, ...   
3178  Omoni Oboli, Kenneth Okonkwo, Kalu Ikeagwu, Uc...   
3210  Hakim Faris, Victoire Du Bois, Patrick d'Assum...   
6112                                    J. Michael Long   
2954  Jyothika, Sathyaraj, Karthi, Showkar Janaki, N...   
8191                                               None   
6545                  Noriko Shinohara, Ushio Shinohara   
6686  Paul Blake, Jeremy Bulloch, John Chapman, Anth...   
1186  Hiroaki Hirata, Yuki Kaji, Asami Seto, Hiroki ...   
3019  Kento Yamazaki, Mackenyu, Mei Nagano, Mamoru M...   
6216  Additi Gupta, Sanjeeda Sheikh, Kiran Karmarkar...   

                                          country          date_added  \
4862                                        Japan        May 15, 2018   
8702  United States, Brazil, India, Uganda, China       July 25, 2018   
3011                                       Canada    January 17, 2020   
3787                                       Canada        May 31, 2019   
624                                          None       June 30, 2021   
8211                                United States      August 7, 2017   
226                                          None     August 25, 2021   
6161                                United States  September 17, 2019   
2022  

In [17]:
df_netflix.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 894.5+ KB


In [18]:
df_netflix.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

Se continua analizando en detalle distintas columnas

Conclusiones del EDA: 

GENERALES (considerando el objetivo y alcance de este proyecto): 

    - Las columnas duration tienen tanto información en minutos como de temporadas

    - Las columnas cast tienen los actores separados por una coma

PARTICULARES

- 1- amazon_prime_titles.csv
    - separdor ','+ encoding utf-8
    - cararcteres especiales [ (30) 
    - valores faltantes 
        - director        2082
        - cast            1233
        - country         8996
        - date_added      9513
        - rating           337
    - 9687 registros 
    - 12 columnas

- 2- disney_plus_titles.csv
    - separdor ','+ encoding utf-8
    - cararcteres especiales [ (NO)  
    - valores faltantes 
        - director        473
        - cast            190
        - country         219
        - date_added        3
        - rating            3
    - 1450 registros 
    - 12 columnas

- 3- hulu_titles.csv
    - separdor ','+ encoding utf-8
    - cararcteres especiales [ (1)  
    - valores faltantes 
        - director        3070
        - cast            3073
        - country         1453
        - date_added        28
        - rating           520
        - duration         479
        - description        4
    - 3090 registros 
    - 12 columnas

- 4- netflix_titles.json
    - valores faltantes 
        - director        2634
        - cast             825
        - country          831
        - date_added        10
        - rating             4
        - duration           3
    - 8807 registros
    - 12 columnas